In [4]:
import numpy
import math
import random

## FUNCTION 1/6 - after06 Median


def median(li):  # compute the median value from a 1D list

    return the median value or upper median if even number of values
    
    example: median([1,3,5,7,9]) returns 5, median([1,3,5,7,9,10]) returns 7

In [9]:
#code block of original code

def median(li):
    n = len(li)
    li.sort()
    if n % 2 == 0:
        return (li[n//2 - 1] + li[n//2]) / 2
    else:
        return li[n//2]

In [10]:
#new, refactored function with name as function2

def median2(li):
    li.sort()
    return li[len(li) // 2]

In [21]:
#use a for loop to create 10 lists of randomly selected numbers to use for testing
tests = [[random.randint(1,100) for i in range(10)]for i in range(10)]
tests
for item in tests:
    print(item)

[76, 54, 20, 27, 52, 23, 88, 7, 95, 8]
[31, 31, 31, 96, 67, 1, 60, 74, 60, 12]
[9, 21, 92, 76, 28, 24, 73, 77, 9, 62]
[21, 14, 43, 38, 51, 45, 98, 3, 90, 54]
[73, 41, 98, 17, 87, 97, 87, 65, 61, 32]
[82, 58, 71, 67, 17, 72, 22, 45, 42, 64]
[25, 14, 79, 55, 38, 81, 45, 90, 45, 71]
[41, 99, 25, 98, 17, 13, 38, 82, 19, 36]
[27, 96, 62, 4, 85, 71, 32, 18, 5, 53]
[83, 9, 15, 24, 99, 22, 10, 92, 64, 6]


In [22]:
%%timeit
results = [median(item) for item in tests]

#timing and testing original function

2.97 µs ± 3.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [23]:
%%timeit
results = [median2(item) for item in tests]

#timing and testing new function

1.84 µs ± 11 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


My new code here provides a much more elegant solution than my original code.

My original code was lengthy and more convoluted than it needed to be. 

My new code accomplishes what my original did, but only in two lines. 

When timed, we can see that the new, refactored code executes in 1.84 microseconds,

while the old code executes in 2.97 microseconds.

## FUNCTION 2/6 - after02 Into

def into(x,y): # function to answer how many times the first number (x) goes into the second(y)

both are integers 

    # return both the number of integer times and the remainder 
    # Example: into(3,10) returns (3,1)

In [34]:
#code block of original code

def into(x,y):
    if x == 0:
        return 'infinite'
    else: 
        return math.floor(y / x), y%x

In [35]:
# new refactored funcrtion with name as function 2

def into2(x,y):
    return(y//x, y%x)

In [36]:
pairs = []

# loop runs 10 times to create 10 pairs
for i in range(10):
    # get two random integers between 1 and 100
    num1 = random.randint(1, 100)
    num2 = random.randint(1, 100)
    # put the pair in the list
    pairs.append((num1, num2))

print(pairs)


[(31, 91), (2, 34), (100, 18), (70, 9), (19, 55), (67, 22), (25, 7), (95, 4), (43, 76), (9, 98)]


In [41]:
%%timeit
results = [into(num1,num2) for item in pairs]
#timing and testing original function

4.18 µs ± 36.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [43]:
%%timeit
results = [into2(num1,num2) for item in pairs]
#timing and testing new function

2.89 µs ± 142 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


My new code eliminates the use of an if-else statement.

It also eliminates the need to specicifically outline edge cases such as when x = 0.

We can see the change in performance of this code, as the new code took 2.89 microseconds to execute, 

while the old code took 4.18 microseconds to execute.

## FUNCTION 3/6 - after04 addevens

def addevens(s):    #add all even integers less than or equal to a passed value (s)

    #return the sum
    # example addevens(5) returns 6 (2 + 4)
    # example addevens(24) returns 156

In [44]:
#code block of original code

def addevens(s):
    k = 0
    for num in range(s+1):
        if num % 2 == 0:
            k += num
    return k

In [45]:
# new refactored code with name as function 2

def addevens2(s):
    k = 0
    for num in range(s+1,2):
        k += num
    return k

In [46]:
tests = [random.randint(1,100) for i in range(10)]

In [47]:
%%timeit
results = [addevens(item) for item in tests]

#timing and testing original function

63.4 µs ± 672 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [48]:
%%timeit
results = [addevens2(item) for item in tests]

#timing and testing new function

3.99 µs ± 58.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


My new code performs drastically better than my old code. 

This is because I institute a step in my range, rather than

asking the code to check the math on if num % 2 == 0 for each number.

The old code executed at 63.4 microseconds, with a 672 nanosecond std dev.

The new code executed at 3.99 microseconds with a 58.6 nanosecond std. dev.

This is a 16x increase in speed

## FUNCTION 4/6 - after04 else1

def else1(n): # given a number, we want to see if it has a factor 5 or less

    #return True if so, False if not, stop checking once you know it is True
    #example else1(23)  returns False as none of the values 2, 3, 4, 5 are integer factors of 23
    #example else1(15) returns True as 3 is a factor of 15


In [49]:
#code block of original code

def else1(n):
    if n%2 == 0:
        return True
    if n%3 == 0:
        return True
    if n%4 == 0:
        return True
    if n%5 == 0:
        return True
    else:
        return False

In [50]:
# new refactored code with name as function 2

def else12(n):
    for x in range(2,6):
        if n % x == 0:
            return True
        else:
            return False

In [54]:
tests = [random.randint(1,100) for i in range(10)]
print(tests)

[36, 61, 94, 80, 42, 26, 65, 52, 22, 22]


In [55]:
%%timeit
results = [else1(item) for item in tests]

#timing and testing original function

2.64 µs ± 94.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [56]:
%%timeit
results = [else12(item) for item in tests]

#timing and testing new function

4.66 µs ± 191 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


After refactoring this function, it turns out my original code was

actually more efficient despite containing more code. 

My old code executed in 2.64 microseconds while 

the new executed in 4.66. 

This goes to show that more lines of code doesn't necessarily mean less efficient.

It's about the content of the code - no loops is more efficient than having a loop.

## FUNCTION 5/6 - after03 bmi

def bmi(weight, height):  #computes a fitness category based on the body mass index (bmi). bmi = weight (lbs.) / height**2 (inches) * 703

    # return the BMI number and class in () on the return line: 
    #   > 30 "obese", 
    #   25-30 "overweight", 
    #   18.5 - 25 "healthy weight", 
    #   16 - 18.5 "underweight", 
    #   < 16 "severely underweight"
    # example bmi(230,73) returns (30.34152749108651, 'obese'), boundary values go in the lower (for example 25 returns 'healthy weight')

In [68]:
#code block of original code

def bmi(weight,height):
    bmi = ((weight) / (height)**2 * 703)
    if bmi > 30:
        return bmi, 'obese'
    elif 25 < bmi <= 30:
        return bmi, 'overweight'
    elif 18.5 < bmi <= 25:
        return bmi, 'healthy weight'
    elif 16 < bmi <= 18.5:
        return bmi, 'underweight'
    elif bmi > 16:
        return bmi, 'severely underweight'

In [69]:
# new refactored code with name as function 2

def bmi2(weight,height):
    bmi = ((weight) / (height)**2 * 703)
    if bmi > 30:
        return bmi, 'obese'
    elif bmi > 25:
        return bmi, 'overweight'
    elif bmi > 18.5:
        return bmi, 'healthy weight'
    elif bmi > 16:
        return bmi, 'underweight'
    elif bmi < 16:
        return bmi, 'severely underweight'

In [70]:
pairs = []

# loop runs 10 times to create 10 pairs
for i in range(10):
    # get two random integers between 1 and 100
    num2 = random.randint(1, 100) #generate random heights between 1-100 inches tall
    num1 = random.randint(1, 500) #generate random weights 1-500 pounds
    # put the pair in the list
    pairs.append((num1, num2))

print(pairs)


[(103, 89), (417, 43), (293, 72), (46, 46), (300, 86), (460, 69), (347, 32), (20, 50), (437, 69), (298, 58)]


In [73]:

%%timeit
results = [bmi(num1,num2) for item in pairs]

#timing and testing original function

8.26 µs ± 165 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [74]:

%%timeit
results = [bmi2(num1,num2) for item in pairs]

#timing and testing new function

8.15 µs ± 129 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


My new code eliminates the need to classify bmi within ranges.

For example, instead of saying any bmi between 18.5 and 25 is healthy,

the new code designates anything above 18.5 as healthy, 

with additional parameters designating >25 as overweight 

and >30 as obese. Once the if statement reaches a valid return, it stops.

This makes the code a tad faster and the code more elegant in nature.

## FUNCTION 6/6 - Assignment1 intersect2d

"intersect2d” accepts 8 numbers representing the opposite corners of 2 axis-aligned 2-dimensional rectangles

(Lower Left and Upper Right). The coordinates are to be passed to the function as follows

I will test with a statement like: intersect2d(0,0,3,3,4,4,5,5)
two opposite corners of the first 2d rectangle as (X,Y) 

coordinates entered as (example values above):

i. LL X (0)
ii. LL Y (0)
iii. UR X (3)
iv. UR Y (3)

two opposite corners for the second 2d rectangle as (X,Y) coordinates entered as:
v. LL X (4)
vi. LL Y (4)
vii. UR X (5)
viii. UR Y (5)

The function should then compute whether or not the 

two rectangles intersect and return Boolean True if they
intersect, or Boolean False if they do not intersect. 

Consider touching to be an intersection.


In [75]:
#code block of original code

def intersect2d(LLX1, LLY1, URX1, URY1, LLX2, LLY2, URX2, URY2):
    
    if URX1 < LLX2:
        return False
        
    elif LLX1 > URX2:
        return False
        
    elif URY1 < LLY2:
        return False
        
    elif LLY1 > URY2:
        return False
        
    else:
        return True

In [109]:
# new refactored code with name as function2

def intersect2d2(LLX1,LLY1,URX1,URY1,LLX2,LLY2,URX2,URY2):
    if URX1 < LLX2 or LLX1 > URX2 or URY1 < LLY2 or LLY2 > URY2:
        return False
    else:
        return True
        
        
    

In [110]:
tests = [[random.randint(1,100) for i in range(8)]for i in range(10)]
print(tests)

[[45, 9, 89, 66, 32, 91, 76, 14], [82, 20, 93, 29, 97, 16, 11, 10], [46, 19, 7, 62, 19, 46, 36, 1], [10, 20, 8, 92, 43, 84, 82, 25], [43, 48, 63, 49, 43, 60, 72, 5], [83, 1, 18, 10, 77, 56, 54, 52], [100, 59, 66, 32, 91, 78, 24, 59], [59, 37, 75, 8, 2, 54, 15, 34], [75, 60, 41, 5, 35, 28, 35, 60], [3, 4, 71, 62, 35, 58, 75, 58]]


In [111]:
%%timeit
results = [intersect2d(*item) for item in tests]

#timing and testing original function

2.95 µs ± 95.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [114]:
%%timeit
results = [intersect2d2(*item) for item in tests]

#timing and testing new function

2.98 µs ± 38.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


When timed, we can see that my old and new code performed very similarly.

However, my new code presents a much more concise and thus elegant solution.

I was able to pare down my code to an if-else statement, when 

originally it was if-elif-elif-elif-else.